relay-server/src/services/processor/state.rs

```diff
@@ -0,0 +1,94 @@
+//! This module contains types and utils used in processor to decribe and support the processing pipeline.
+//!
+
+use crate::services::processor::ProcessEnvelopeState;
+
+macro_rules! state_type {
+    ($name:ident) => {
+        pub struct $name<'a, G>(ProcessEnvelopeState<'a, G>);
```

Comment:
To allow running the same function from different groups, the inner would need to be generic and to enforce certain properties we'd have to use traits, is that still something you want to get to or do you have a new approach in mind?